granulate_utils/linux/cgroups/cgroup.py

```diff
@@ -78,7 +95,99 @@ def find_v2_hierarchy(resolve_host_root_links: bool = True) -> Optional[str]:
     return path
 
 
-def get_cgroup_mount(controller: str, resolve_host_root_links: bool = True) -> Optional[str]:
+CGROUP_PROCS_FILE = "cgroup.procs"
+
+
+class CgroupCore:
+    path: Path
+
+    def __init__(self, path: Path):
+        self.path = path
+
+    def _get_parent_cgroup(self, cgroup_name: str) -> Optional[Path]:
+        cgroup_index = self.path.parts.index(cgroup_name)
+        if cgroup_index != -1:
+            return Path(*self.path.parts[: cgroup_index + 1])
+        return None
+
+    def _create_subcgroup(self, cgroup_name: str) -> Optional[Path]:
+        if self.has_parent_cgroup(cgroup_name):
+            new_cgroup_path = self._get_parent_cgroup(cgroup_name)
+        else:
+            new_cgroup_path = self.path / cgroup_name
+            new_cgroup_path.mkdir(exist_ok=True)
+
+        return new_cgroup_path
+
+    def has_parent_cgroup(self, cgroup_name: str) -> bool:
+        """
+        Check if cgroup has the given cgroup_name in its hierarchy
+        :param cgroup_name: the cgroup name
+        :return: bool indicating if cgroup is in hierarchy
+        """
+        controller_path_parts = self.path.parts
+        return cgroup_name in controller_path_parts
+
+    def get_pids_in_cgroup(self) -> set[int]:
+        return {int(proc) for proc in self.read_from_interface_file(CGROUP_PROCS_FILE).split()}
+
+    def assign_process_to_cgroup(self, pid: int = 0) -> None:
+        if not self.path.exists():
+            raise FileNotFoundError("Cgroup doesn't exist")
```

Comment:
1. Create a specific exception type for it. Personally I dislike raising builtin exceptions for non-builtin cases, especially when the message is modified and the meaning is different
2. Don't check and raise then run the opreation. Just run the operation and let the exception pass up, OR catch and re-raise appropriately. Checking-raising-writing is racy (what if file goes down between the check and the write?) and error prone in general (what if the error can be raised in cases you are not checking before?)

Reply:
yeah you're right it is racey, I'm removing it